src/main/java/io/debezium/connector/vitess/VitessConnector.java

```diff
@@ -113,18 +113,53 @@ public List<Map<String, String>> taskConfigs(int maxTasks, List<String> shards)
                                 + "Otherwise please reset the offset.storage.task.key.gen config to its original value",
                         prevNumTasks, tasks));
             }
+            // Check the task offsets persisted from previous gen, we expect the offsets are saved
             Map<String, String> prevGtidsPerShard = gen > 0 ? getGtidPerShardFromStorage(prevNumTasks, gen - 1, true) : null;
             LOGGER.info("Previous gtids Per shard: {}", prevGtidsPerShard);
-            if (prevGtidsPerShard != null && prevGtidsPerShard.size() != shards.size()) {
-                throw new IllegalArgumentException(String.format(
-                        "Different number of shards between offset storage: %s and current vitess shards: %s. ",
-                        prevGtidsPerShard, shards));
+            if (prevGtidsPerShard != null && prevGtidsPerShard.size() != currentShards.size()) {
+                LOGGER.warn("Some shards for the previous generation {} are not persisted.  Expected shards: {}",
+                        prevGtidsPerShard.keySet(), currentShards);
             }
             final String keyspace = connectorConfig.getKeyspace();
+            // Check the task offsets for the current gen, the offset might not be persisted if this gen just turned on
             Map<String, String> gtidsPerShard = getGtidPerShardFromStorage(tasks, gen, false);
-            if (gtidsPerShard != null && gtidsPerShard.size() != shards.size()) {
+            if (gtidsPerShard != null && gtidsPerShard.size() != currentShards.size()) {
```

Comment:
@HenryCaiHaiying The code lGTM.One question though. Is it possible that while Debezium is down that there will be split and later on merge so the shards are different but the count is the same?

Reply:
In theory this can happen but in practice, the chances are very low due to:
1. debezium down is not likely going over multiple hours otherwise this realtime change capture pipeline downstream users will be in trouble;  And if debezium down for too long, it will have problems on reconnecting back to Vitess DB since the gtid position it stored previously might already be removed from Vitess side due to Vitess data retention policy;
2. Vitess shard split is a somewhat likely event but merge back is a very rare event.  In our company (Slack) of using Vitess for more than 3 years, there is probably only once or twice merge back event.
3. Vitess shard split (or merge back) is usually a planned event executed manually.  The chance of vitess shards split and then merge back during the debezium down time without notice is very unlikely.

Reply:
But if you feel strongly we should check against this situation, I can also change the comparison code to do that.

Reply:
Even if it is rare, can't it lead into a hard-to-identify issue? If yes then I'd prefer to hedge against it too.

Reply:
OK, I've added a new method hasSameShards() to do element-wise content comparison (instead of just relying on the size check), please take a look @jpechane

Reply:
Ping @jpechane ...